src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs

```diff
@@ -275,5 +327,35 @@ private void AssignTimeout(long ticks, TimeoutAction timeoutAction)
             // Add Heartbeat.Interval since this can be called right before the next heartbeat.
             Interlocked.Exchange(ref _timeoutTimestamp, _lastTimestamp + ticks + Heartbeat.Interval.Ticks);
         }
+
+        public void StartTimingReads()
+        {
+            Interlocked.Exchange(ref _readTimingElapsed, 0);
+            Interlocked.Exchange(ref _bytesRead, 0);
+            Interlocked.Exchange(ref _readTimingStartTicks, _lastTimestamp);
```

Comment:
Why are these Interlocked.Exchange's helping? If we're worried about concurrent access/modification of these fields, wouldn't these fields being momentarily inconsistent with each other be a problem?

You can't safely do `+=` operations to fields that are reset concurrently even with Interlocked.Exchange, so `_readTimingElapsed += timestamp - _lastTimestamp;` seems unsafe. I don't think setting `_timingReads` to true last saves you here.